lib/std/http/Client.zig

```diff
@@ -657,7 +657,7 @@ pub const Request = struct {
         };
     }
 
-    pub const SendError = Connection.WriteError || error{ InvalidContentLength, UnsupportedTransferEncoding };
+    pub const SendError = Allocator.Error || Connection.WriteError || error{ InvalidContentLength, UnsupportedTransferEncoding };
```

Comment:
I don't think there should be a possible `OutOfMemory` in send().

Look at what a `Connection` has:

```zig
    pub const buffer_size = std.crypto.tls.max_ciphertext_record_len;
// ...
    read_start: BufferSize = 0,
    read_end: BufferSize = 0,
    write_end: BufferSize = 0,
    read_buf: [buffer_size]u8 = undefined,
    write_buf: [buffer_size]u8 = undefined,
```

This buffer is enormous. It can definitely hold any formatted string that needs to be allocated. And in the rare case that the buffer gets full, the full buffer can be sent to the OS and then the data can be added to the freshly empty buffer.

The writer implementation on Connection already does the right thing. It just needs to be used for formatted printing rather than external heap allocation.

Reply:
std.base64 does not have a streaming encoder, so at the moment both the input (to merge the user and password) and output buffers need to be allocated before they can be written.

Reply:
On the output buffer side, you can obtain an output buffer from the Connection `write_buf`. Just write to the buffer.

For the input buffer, use the stack.

[popular servers chose 255 for this limit](https://httpd.apache.org/docs/2.4/programs/htpasswd.html#restrictions)

Reply:
I agree that if there were a streaming base64 encoder then we wouldn't have to scrounge for memory at all, simply stream each part to the final writer, which is the main reason I didn't spend time looking for a better solution.

Reply:
I don't think "scrounging" is a fair characterization. The output buffer is the correct place to write memory to, and using the stack space is hardly unusual, especially since there is most certainly an upper bound on the size of the input data.

Besides, a streaming encoder:
* would be slightly more bloated
* would be slightly slower
* as you noted, isn't currently implemented